### DISCUSSION TITLE
Question about compare.

### DISCUSSION THREAD

**Discussion by acb764se** [ImageMagick/ImageMagick#3692] [Help]
Hi, I am confused about the fuzz factor.
All is clear when I compare 2 images from the CLI:
compare -matric AE -fuzz 10% pic1.jpg pic2.jpg /dev/null/aa.jpg

but when I do it from my code, I find that the compare function has no "fuzz" input, I have to do instead:
....
  prevPicture->colorFuzz(10*QuantumRange/100);
  nextPicture->colorFuzz(10*QuantumRange/100);
  double test = prevPicture->compare(*nextPicture,AbsoluteErrorMetric);
....

it looks like "fuzz" is a property of any one image.
What if I had used 2 different values in the 2 first lines above? Which one would have been used to tell if 2 pixels in the same position are equal or not??

I am using Magick++ version 6. Thank you!

**Comment by snibgo** [comment]
Tracing through the code takes me to GetFuzzyColorDistance() in color-private.h, and the answer seems to be: "both". The two fuzzes are multiplied.

But I suggest you experiment to confirm this is actually what happens.

**Comment by acb764se** [comment]
Hi, thanks for your interest!
I tried looking into the code myself, but it is obscured way beyond my (limited, undoubtedly) capacity.
So i tried the way of experimental informatics that you suggested.

I used a pair of very similar pictures from one of my security cameras, taken 10s apart, 1920x1080, sRGB,
so they have 2,073,600 pixels times 3 colors each, so 6,220,800 8-bit numbers in all. I got this table! YOW!

fuzz factor 1       fuzz factor 2      different pixels (or pixel-and-color?)
10%                   10%                  47,705
5%                      10%                 47,705
10%                   5%                     47,705
5%                     5%                     308,765
def.                    def.                    1,848,760              <-this "default" was obtained without setting any fuzz.
2%                     2%                    922,862
1%                     1%                     1,503,560
0%                     0%                     1,848,760

So, I conclude: 
1) there seems to be a default, this default is 0%, which is logical, and I see only 1,8M differences because my 2 pictures are very similar.
2) for this same reason, lowering only one fuzz factor seems to have no effect, the test is dominated by the bigger one.

The only way that I can make sense of each picture having its own fuzz factor is that -by setting one- each color
is turned for comparison (and possibly other) purposes into an interval, and this interval has a right 
to be different for every picture.
For example, imagine that in my pictures, in a corresponding position, I have pixels with Red component 100 and 140.

picture 1: 100 +=30% becomes an interval from 70 to 130
picture 2: 140 +=5%  becomes an interval from 133 to 147 ---------------------> these pixels are DIFFERENT!

but, exchanging fuzz factors,

picture 1: 100 +=5% becomes an interval from 95 to 105
picture 2: 140 +=30%  becomes an interval from 98 to 182 ---------------------> these pixels are EQUALS!

because in the second case the intervals overlap, while in the first one they do not.

But maybe the algorithm behind the scenes is not this at all.

Thank you all if you have felt an interest in this!

**Comment by snibgo** [comment]
Sorry, I was wrong. GetFuzzyColorDistance() in color-private.h takes the _maximum_ of the two fuzzes. It does not multiply them. So (10%,10%) and (5%,10%) and (10%,5%) should return the same result.
